File: problems-originals/0701-0800/0743_network-delay-time/solution_bellman_ford.cpp
class Solution {
  public:
    int networkDelayTime(vector<vector<int>> &times, int n, int k) {
        const int INF = 1e8;
        vector<int> dist(n + 1, INF);
        dist[k] = 0;
        // Each round extends shortest paths by one edge, so n-1 rounds suffice.
        for (int round = 0; round < n - 1; ++round) {
            bool changed = false;
            for (const auto &t : times) {
                // The dist[u] finite guard keeps INF + w from overflowing.
                if (dist[t[0]] < INF && dist[t[0]] + t[2] < dist[t[1]]) {
                    dist[t[1]] = dist[t[0]] + t[2];
                    changed = true;
                }
            }
            // A round that relaxes nothing means the distances are final.
            if (!changed)
                break;
        }
        int best = 0;
        for (int i = 1; i <= n; ++i) {
            // Anything still at INF is unreachable from k.
            if (dist[i] >= INF)
                return -1;
            best = max(best, dist[i]);
        }
        return best;
    }
};

File: problems/0701-0800/0743_last-to-hear-the-signal/solution_bellman_ford.go
func lastToHear(edges [][]int, n int, k int) int {
	const inf = 100000000
	dist := make([]int, n+1)
	for i := range dist {
		dist[i] = inf
	}
	dist[k] = 0
	// Each round extends shortest paths by one edge, so n-1 rounds suffice.
	for round := 0; round < n-1; round++ {
		changed := false
		for _, t := range edges {
			// The dist[u] finite guard keeps inf + w from overflowing.
			if dist[t[0]] < inf && dist[t[0]]+t[2] < dist[t[1]] {
				dist[t[1]] = dist[t[0]] + t[2]
				changed = true
			}
		}
		// A round that relaxes nothing means the distances are final.
		if !changed {
			break
		}
	}
	best := -1
	for i := 1; i <= n; i++ {
		// Anything still at inf is unreachable from k.
		if dist[i] >= inf {
			return -1
		}
		if dist[i] > best {
			best = dist[i]
		}
	}
	return best
}
